[Form] DateTimeImmutable norm data in DateTime form types

vudaltsov opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | symfony/symfony-docs#... <!--highly recommended for new features-->

This PR suggests a way to use `\DateTimeImmutable` objects for norm data in `DateTimeType`, `DateType` and `TimeType`.

A complete BC layer for transformers is provided. All protected calls are also taken into account.

If this PR gets merged, we can then easily add `input=datetimeimmutable` with no model transformers to support `\DateTimeImmutable` model data.

vudaltsov committed:
DateTimeImmutable norm data in DateTime form types

vudaltsov committed:
Fixed the codestyle issues

vudaltsov commented:
ping @xabbuh , @stof, @HeahDude

Simperfit commented on src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeImmutableToArrayTransformer.php:
```diff
@@ -0,0 +1,184 @@
+class DateTimeImmutableToArrayTransformer extends BaseDateTimeTransformer
+{
+    private $pad;
+
```
no need for this empty line

vudaltsov replied:
thanx, done

Simperfit commented on src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeImmutableToLocalizedStringTransformer.php:
```diff
@@ -0,0 +1,202 @@
+class DateTimeImmutableToLocalizedStringTransformer extends BaseDateTimeTransformer
+{
+    private $dateFormat;
+
```
same

Simperfit reviewed (changes requested):
You need to add tests for all the new classes you created.

vudaltsov committed:
Removed blank lines in property declarations.

vudaltsov committed:
Tests for transformers

vudaltsov commented:
Added tests.

stof commented:
I'm not sure about the BC impact of changing the type of the norm data. I suggest adding the new input type first, without changing the norm data.

vudaltsov commented:
[Here](https://github.com/symfony/symfony/pull/19889#issuecomment-283394385) @HeahDude posted:

> If a new major version of Symfony bumps the requirement to PHP 5.5+, the normalized data could then be switched to the DateTimeImmutable class to simplify the logic (which is currently cloning the DateTime to avoid modifying the original one.

That's what I tried to do in my PR. My change is pretty straightforward: I copied and renamed transformers and their tests and removed `clone`.

If we really want to make norm data immutable than I think we should do it first. If we don't - than yes, just add `input=datetimimmutable` and a corresponding transformer.

But with what I suggest immutable dates will be supported in the first place. Only `DateTime`s will be converted. In this way immutable norm data helps to follow the trend of replacing `DateTime` with `DateTimeImmutable`.

stof commented:
you need to add the new allowed value for `input` option. Otherwise, immutable datetime are **not** supported (they are used as norm data, but cannot be used as model data, which is the requested feature)

vudaltsov commented:
Sure! I planned to do that in a separate PR. Because it's gonna have a couple of new tests.
In this PR I just provide a ground for that by changing the norm data.

If you wish, I can add commits right here.

vudaltsov commented:
@Simperfit, could you please review my tests?

Simperfit commented on src/Symfony/Component/Form/Tests/Extension/Core/DataTransformer/DateTimeImmutableToTimestampTransformerTest.php:
```diff
@@ -0,0 +1,103 @@
+    {
+        $reverseTransformer = new DateTimeImmutableToTimestampTransformer();
+
+        $this->{method_exists($this, $_ = 'expectException') ? $_ : 'setExpectedException'}('Symfony\Component\Form\Exception\TransformationFailedException');
```
maybe you should add a new var for this.

Simperfit reviewed (commented):
The tests looks good to me,

vudaltsov commented:
Closed in favor of https://github.com/symfony/symfony/pull/25273.
